Add caching

enescualexandru opened the pull request.

enescualexandru committed:
Added project structure and basic implementation

enescualexandru committed:
Add caching

enescualexandru committed:
Addressed code review comments

enescualexandru committed:
Fixed rebasing conflicts

- renamed/added packages
- refactored class/interface inheritance
- amended readme file, added caching

enescualexandru committed:
Added tests for verifying the caching layer; added small improvements to the code

TheBestPessimist commented on README.md:
```diff
@@ -68,6 +68,9 @@ Meaning that for eg. `google.com` will be accepted, while `$#@le.com` will not.
 is started on debug mode, the debug port is `5005`. The postgres db can be reached at port `5432`, with the dbname, user
 and password specified in the docker-compose file.
+
+**NOTE FOR CACHING:**  Redis cli can be accessed by `docker exec -it redis redis-cli`, then `auth {password}`(password in the docker-compose file).
```
super very nice documentation!

TheBestPessimist commented on docker-compose.yml:
```diff
@@ -30,5 +34,19 @@ services:
\ No newline at end of file
+  postgres-data:
+  redis-data:
+  redis-conf:
```
no newline at the end of file
![image](https://user-images.githubusercontent.com/4482210/133456234-5c9222b4-ba80-489d-a1e6-a657929a4be0.png)

TheBestPessimist commented on docker-compose.yml:
```diff
@@ -30,5 +34,19 @@ services:
       - POSTGRES_PASSWORD=postgres
       - POSTGRES_DB=url-shortener
 
+  redis:
```
dont forget about `depends on redis`

TheBestPessimist commented on src/main/kotlin/com/shortener/data/cache/RedisCacheConfiguration.kt:
```diff
@@ -0,0 +1,23 @@
+class RedisCacheConfiguration {
+
+    @Bean
+    fun redisTemplateUrlEntry(connectionFactory: RedisConnectionFactory?): RedisTemplate<String, UrlEntry> {
```
why do we want to receive a nullable `connectionFactory` ?

enescualexandru committed:
Addressed code review comments:

- docker-compose: added upstream dependency to redis container; added empty line at eof
- ensured connectionFactory is not nullable in the redis template bean

TheBestPessimist commented on src/main/kotlin/com/shortener/data/cache/UrlEntryRepositoryCacheImpl.kt:
```diff
@@ -0,0 +1,45 @@
+
+    @Autowired
+    @Qualifier("urlEntryRepository")
+    private lateinit var urlEntryRepository: UrlEntryRepositoryBase
```
Please dont use `@Autowired` on fields in `main`. I can accept that in `test`, but not in `main`.

Instead use constructor parameters.

enescualexandru replied:
Will correct it. I missed it again, need to break the old habit

TheBestPessimist commented on src/test/kotlin/com/shortener/TestUtils.kt:
```diff
@@ -0,0 +1,28 @@
+const val DEFAULT_SEQUENCE = "a1b2"
+const val DEFAULT_EXPIRED = false
+
+class TestUtils {
```
This class acts as a singleton containing (probably) only static functions. In this case, you can make `TestUtils` an `object` (another concept to read about).

By having `object TestUtils`, you no longer need a `companion object`. Each of the functions inside are static.

You may also consider removing `TestUtils` completely as it has a single function and leave it as a `top level function`, similar to the `const val`s above (`top level variable`s).

enescualexandru replied:
Noted, thanks for the tip. I would prefer to keep the `TestUtils` as a separate class(object), considering the functions additions in the future(create test users etc.)

TheBestPessimist commented on src/test/kotlin/com/shortener/TestUtils.kt:
```diff
@@ -0,0 +1,28 @@
+            expired: Boolean = DEFAULT_EXPIRED
+        ): UrlEntry =
+            UrlEntry(longUrl).apply {
+                encodedSequence = EncodedSequence().apply { sequence = encodedSequenceStr }
```
I have seen this pattern `EncodedSequence().apply { sequence =` in code a lot of times and it started to bother me. Maybe you can add `sequence` as the first constructor param (or in a secondary constructor) instead.

enescualexandru replied:
Will do.

TheBestPessimist commented on src/test/kotlin/com/shortener/data/cache/UrlEntryRepositoryCacheImplTests.kt:
```diff
@@ -0,0 +1,88 @@
+    fun getValueFromByteArray(value: ByteArray): UrlEntry? =
+        redisTemplate.valueSerializer.deserialize(value) as UrlEntry
+
+    fun getUrlEntryFromCache(sequence: String?): UrlEntry? =
```
can these helper functions be private?

enescualexandru replied:
Yes, they should be. I will correct them

TheBestPessimist commented on src/main/resources/application.properties:
```diff
@@ -1,2 +1,6 @@
+
+spring.cache.redis.cache-null-values=false
+# 24 hours
+spring.cache.redis.time-to-live=1440000
```
if we dont cache by default, how do we use redis?

isnt 
```properties
spring.cache.type=redis
```
missing?

enescualexandru replied:
Yet another spring boot thing. I have `spring-boot-starter-data-redis` as a dependency, and I guess spring detects it as the provider. And caching is declaratively enabled in the `RedisCacheConfiguration` class. In the [official spring boot documentation ](https://docs.spring.io/spring-boot/docs/2.1.6.RELEASE/reference/html/boot-features-caching.html#boot-features-caching-provider)(see the first remark in that section) is advised to set this property when you want to disable the caching(which I am doing for testing - enabling or disabling)

TheBestPessimist commented on src/main/kotlin/com/shortener/service/UrlServiceImpl.kt:
```diff
@@ -24,7 +25,8 @@ private val baseUrl = ServletUriComponentsBuilder.fromCurrentContextPath().build
     private val idEncoder: IdEncoder,
-    private val urlValidator: UrlValidator
+    private val urlValidator: UrlValidator,
+    private val urlEntryRepositoryCacheImpl: UrlEntryRepositoryBase
```
architectural question: why does `UrlServiceImpl` care if the repository uses cache or not? It is reflected in the name of the variable.

What bothers me more though, is that now we ahve 2 repositories: one with cache, another without. during autowiring someone might ask for the wrong repository.

enescualexandru replied:
Probably the implementation is not so coherent as I would have wanted to be. The intention here was to have a simple jpa repo and a caching decorator around it(having the same common interface, `UrlEntryRepositoryBase`), allowing the clients to use either of the two. Why there are still two repos ? Some methods return data which I didn't want it cached. Probably I should've use some kind of factory to generate the correct repo for the client

TheBestPessimist commented on src/main/kotlin/com/shortener/service/UrlServiceImpl.kt:
```diff
@@ -24,7 +25,8 @@ private val baseUrl = ServletUriComponentsBuilder.fromCurrentContextPath().build
 class UrlServiceImpl(
     private val urlEntryRepository: UrlEntryRepository,
```
why do we use both `UrlEntryRepository` and `UrlEntryRepositoryBase`? 

`UrlEntryRepository` is a child of `UrlEntryRepositoryBase`, so this looks awkward

enescualexandru replied:
Further to the answer from the question above: `UrlEntryRepositoryBase` is defining the methods which are cached(well, I need to rename it)

TheBestPessimist commented on src/main/kotlin/com/shortener/data/cache/UrlEntryRepositoryCacheImpl.kt:
```diff
@@ -0,0 +1,45 @@
+const val CACHE_NAME_URL_ENTRY = "url-entry-cache"
+
+@Component
+class UrlEntryRepositoryCacheImpl : UrlEntryRepositoryBase {
```
Architectural q

1. Why do we cache spring entities? afaiu caches should hold \*actually useful\* data, and an `Entity` is anything but that. It is an oversized monstrosity (contains so much JPA/Spring AOP bloat it scares me) from the POV of a cache.
2. (i never worked with redis + spring, so bear with me) Why is the cache another `Repository`? Is this how Spring best practices suggests we implement caching? Is this how caching is supposed to work? I would like to hear your thoughts here.
3. If this is a `Repository`, why is it annotated with `@Component`?
4. Are we supposed to never use `UrlEntryRepository` again, since we will potentially be missing on caching?


When i imagined caching, i thought of a separate service which deals with redis as a key-value kind of thing, and which is used by `UrlService`

enescualexandru replied:
1. Not sure what do you mean by that. Every entry in the cache has a key(I'm using the `encoded sequence` for that) and a value(which is a an `UrlEntry`, serialized). On cache hit, that value is deserialized back to an UrlEntry. So what data is not useful ?
2. I also never worked before with redis, and definitely didn't handled the caching mechanism in a declarative way, through annotations. Didn't considered this class as solely 'the cache' . Is the decorator which holds and uses the db repo, when needed, and adds the caching abstraction on it. Couldn't find an official spring documentation for caching best practices.
3. I am thinking of 'Repository' as a place to store data, like db/cache etc. If you mean `@Repository`: I don't think is even needed. From spring documentation, about `@Repository`:  this annotation also serves as a specialization of @Component, **allowing for implementation classes to be autodetected through classpath scanning** - not the case here. 
4. We are supposed to use both cached/not cached versions

Answer to the last remark: maybe so, but then the UrlService needs to decide by it's own when to cache or not and implement the logic for that

enescualexandru committed:
Addressed code review comments:

- code improvements
- fixed util functions from the cache test class
- added one more test for caching
- added the `sequence` property in the constructor(EncodedSequence) class
- TestUtils transformed to object

enescualexandru committed:
Addressed code review comments:

- refactored the caching layer

enescualexandru committed:
Addressed code review comments:

- refactored the caching layer, part 2

enescualexandru committed:
Update pom.xml
